src/app/home/recent-spaces-widget/recent-spaces-widget.component.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  Component,
+  ErrorHandler,
+  Input,
+  OnInit
+} from '@angular/core';
+import { Router } from '@angular/router';
+import { uniqWith } from 'lodash';
+import {
+  Observable,
+  ReplaySubject,
+  Subject
+} from 'rxjs';
+
+import {
+  Broadcaster,
+  Logger
+} from 'ngx-base';
+import {
+  Context,
+  Contexts,
+  Space,
+  Spaces,
+  SpaceService
+} from 'ngx-fabric8-wit';
+
+@Component({
+  selector: 'fabric8-recent-spaces-widget',
+  templateUrl: './recent-spaces-widget.component.html'
+})
+export class RecentSpacesWidget implements OnInit {
+
+  @Input() cardSizeClass: string;
+
+  readonly userHasSpaces: Subject<boolean> = new ReplaySubject<boolean>(1);
+  readonly recentSpaces: Observable<Space[]>;
+
+  constructor(
+    private contexts: Contexts,
+    private spaces: Spaces,
+    private spaceService: SpaceService,
+    private router: Router,
+    private errorHandler: ErrorHandler,
+    private broadcaster: Broadcaster,
+    private logger: Logger
+  ) {
+    this.recentSpaces = this.spaces.recent
+      .map((spaces: Space[]): Space[] => uniqWith(spaces, RecentSpacesWidget.compareSpaces));
+  }
+
+  ngOnInit(): void {
+    (this.router.url.startsWith('/_home') ? this.contexts.default : this.contexts.current)
+      .first()
+      .flatMap((context: Context): Observable<Context> => {
+        if (context && context.user) {
+          return Observable.of(context);
+        } else {
+          return Observable.throw('Failed to retrieve list of spaces owned by user');
+        }
+      })
+      .flatMap((context: Context): Observable<boolean> => this.spaceService
+        .getSpacesByUser(context.user.attributes.username)
+        .map((spaces: Space[]): boolean => spaces.length > 0)
```

Comment:
The `uniqWith` is the actual fix for the duplicate spaces issue, I tried to mark this by using the "fix" verb on that specific commit only but this definitely isn't very visible.

I can certainly add a comment here, but you're right that fixing the `Spaces` to not emit duplicates to begin with would be better, and would remove that logic from this component so no need for a comment. I was avoiding touching the `Spaces` (really the `SpacesService` implementation) because it does touch so many places and is consumed by so many other spaces and components, and I have no idea if any of those might for some reason rely on the fact that it does emit duplicates.

Reply:
I'm fine with not touching SpacesService, but could you please log an issue against the REST API instead? The API should provide unique spaces, not duplicates -- the UI shouldn't be forced to prune them out.

Reply:
It's actually not backed by a REST API:

https://github.com/fabric8-ui/fabric8-ui/blob/5e2074265445165adf27759fb4d42a7296b01c50/src/app/shared/spaces.service.ts#L35

There is a bit of logic there that's meant to eliminate duplicates, but clearly it's not working as intended since there's a bug report showing exactly that happening. I'll open a separate PR to fix the duplication logic there and remove the uniqWith from this one.

Reply:
Sounds good to me